### ISSUE TITLE
Is there a lazy and parallel version of describe() ?

### ISSUE THREAD

**Issue by zacchiro** [pola-rs/polars#8434]
### Research

- [X] I have searched the above polars tags on Stack Overflow for similar questions.

- [ ] I have asked my usage related question on [Stack Overflow](https://stackoverflow.com).

### Question about Polars

Unless I am missing something, there is no `describe()` method on lazy data frames, is there a reason for that?
I'm working with a large CSV: ~9 GiB, with a few billion of columns containing only numerical values, which does fit in memory after conversion to i64, but only a medium-size server.
It should be possible to implement describe() on lazy data frames, right? Or am I missing something?

**Comment by alexander-beedie** [comment]
There isn't, but with a little reworking there could be... I'll take a look.

**Comment by alexander-beedie** [comment]
Ok... so the issue is that in order to generate those statistics you are essentially materialising the frame anyway, so it's better that you're explicitly aware of that by having to call `collect()` beforehand. While looking into it I was able to get some large speedups by reworking it to actively parallelise all of the metrics calculation, so there is that... ;)

**Comment by zacchiro** [comment]
Yeah, I was about to say that there were in fact two issues there: lazyness and parallelism. (My bad for covering only one in the initial question.)

I'm going to try again with the new parallel version, because with the polars pip release as of yesterday, on my dataset Polars' `collect()` seemed slower than Pandas', which felt weird.
Let me know if you want me to post results here.

**Comment by alexander-beedie** [comment]
> Let me know if you want me to post results here.

Definitely! The performance patch just merged, so will be available soon in `0.17.8` 👍

**Comment by ritchie46** [comment]
> I'm going to try again with the new parallel version, because with the polars pip release as of yesterday, on my dataset Polars' collect() seemed slower than Pandas', which felt weird.

Slower than what in pandas? As `collect` includes everything -> IO/ compute / materialization.

**Comment by zacchiro** [comment]
> Slower than what in pandas? As `collect` includes everything -> IO/ compute / materialization.

But I wasn't using `collect()`, precisely due to the lack of a lazy version of the method. So I was comparing a `.describe()` call on either side, *after* (eager) loading.

Anyway, it's pointless to discuss my old benchmark now. I'll do a new one as soon as 0.17.8 hits PyPI.

**Comment by ritchie46** [comment]
Ah, right. :)

Yep, we did not optimize the `describe` until now. :)

Not all of our codebase has been watched with optimization eyes (yet). Let me know how it performs once the release is issued (this afternoon).

**Comment by zacchiro** [comment]
Oh well, I've written my benchmark snippets and tested them on the versions of pandas/polars I was using yesterday, so I might as well post the results now (will update them with the newer release later).

In short: memory usage is about the same between pandas 1.1.5 and polars 0.17.6, but pandas was faster in both reading (2m6s vs 2m30s, *probably* due to rechunk, which I didn't tune) and describing (1m26s vs 3m14s). Gory details below.

Dataset
=======

TL;DR: single column data frame, containing 3 billion integers

	$ du -sh --apparent-size revision-message-length--2023-04-17.txt
	8,8G    revision-message-length--2023-04-17.txt

	$ head -n 3 revision-message-length--2023-04-17.txt
	47
	21
	67

	$ wc -l revision-message-length--2023-04-17.txt
	3069904228 revision-message-length--2023-04-17.txt


Before Polars optimization (discussed in https://github.com/pola-rs/polars/issues/8434 )
==========================

Versions
--------

    polars            0.17.6
    python3-pandas    1.1.5+dfsg-2


Pandas
------

	$ ipython3
	In [2]: # %load describe_commit_length__pandas.py
	   ...: import pandas as pd
	   ...: 
	   ...: %time df = pd.read_csv("revision-message-length--2023-04-17.txt", names=["length"])
	   ...: %time df.info()
	   ...: %time df.describe(percentiles=[.25, .5, .75, .9, .99, .999])
	   ...: 
	CPU times: user 1min 47s, sys: 19.1 s, total: 2min 6s
	Wall time: 2min 6s


	<class 'pandas.core.frame.DataFrame'>
	RangeIndex: 3069903970 entries, 0 to 3069903969
	Data columns (total 1 columns):
	 #   Column  Dtype
	---  ------  -----
	 0   length  int64
	dtypes: int64(1)
	memory usage: 22.9 GB
	CPU times: user 0 ns, sys: 5.76 ms, total: 5.76 ms
	Wall time: 7.23 ms


	CPU times: user 1min 12s, sys: 13.4 s, total: 1min 26s
	Wall time: 1min 26s
	Out[2]: 
				 length
	count  3.069904e+09
	mean   9.263153e+01
	std    9.939477e+03
	min    0.000000e+00
	25%    1.600000e+01
	50%    3.200000e+01
	75%    6.800000e+01
	90%    1.650000e+02
	99%    7.570000e+02
	99.9%  2.938000e+03
	max    1.060886e+08


Polars
------

	$ ipython3
	In [2]: # %load describe_commit_length__polars.py
	   ...: import polars as pl
	   ...: 
	   ...: %time df = pl.read_csv("revision-message-length--2023-04-17.txt", has_header=False)
	   ...: %time df.rename({"column_1": "length"})
	   ...: %time df.estimated_size("gb")
	   ...: %time df.describe(percentiles=[.25, .5, .75, .9, .99, .999])
	   ...: 
	CPU times: user 2min 30s, sys: 32.4 s, total: 3min 2s
	Wall time: 10.6 s
	

	CPU times: user 1.9 ms, sys: 409 µs, total: 2.31 ms
	Wall time: 620 ms


	CPU times: user 43 µs, sys: 0 ns, total: 43 µs
	Wall time: 53.2 µs
    Out[4]: 23.229952764697373


	CPU times: user 8min 24s, sys: 36.3 s, total: 9min 1s
	Wall time: 3min 14s
	Out[2]: 
	shape: (13, 2)
	┌────────────┬─────────────┐
	│ describe   ┆ column_1    │
	│ ---        ┆ ---         │
	│ str        ┆ f64         │
	╞════════════╪═════════════╡
	│ count      ┆ 3.0699e9    │
	│ null_count ┆ 258.0       │
	│ mean       ┆ 92.631532   │
	│ std        ┆ 9939.477111 │
	│ …          ┆ …           │
	│ 75%        ┆ 68.0        │
	│ 90%        ┆ 165.0       │
	│ 99%        ┆ 757.0       │
	│ 100%       ┆ 2938.0      │
	└────────────┴─────────────┘

**Comment by ritchie46** [comment]
> In short: memory usage is about the same between pandas 1.1.5 and polars 0.17.6, but pandas was faster in both reading (2m6s vs 2m30s, probably due to rechunk, 

You seem to compare wall time vs cpu time? Polars is finished in 10s wall time, vs pandas in 2m wall time. CPU times is higher due to doing work in parallel. Polars will almost always have longer CPU time as it does more work (in parallel).

**Comment by zacchiro** [comment]
Ah, yeah, totally for reading CSV, my bad. (I'll update the summary for that part.)
I was more interested/concerned by the `describe()` time, but that is indeed what has the most to gain by the feature you just implemented.

**Comment by zacchiro** [comment]
Just tried it with Polars 0.17.8, here is the result of the last `describe()` step:
```
CPU times: user 12min 10s, sys: 56.8 s, total: 13min 7s
Wall time: 35.4 s
```
This is a server with a lot of cores, where (sloppily) monitoring htop I have noticed a peak of 54 CPUs being used at the same time.

Quite impressive! This is now ~6x faster than before and ~3x faster than Pandas.
Kudos and thanks a lot :-)

**Comment by alexander-beedie** [comment]
> Quite impressive! This is now ~6x faster than before and ~3x faster than Pandas. Kudos and thanks a lot :-)

That's great to hear - thanks for bringing it to our attention. Given the scale of the performance gains I think I'll close this out, as we don't intend to add a LazyFrame variant (given the earlier explanation). Hopefully that makes sense?
